### PR TITLE
various node:buffer fixes

### PR DESCRIPTION
**Pull request by nektro**
individual commits contain fixes and new tests

### CODE REVIEW

**Comment by robobun** [comment]
@nektro, your commit 0494ba2830546d101b5f627f23c9d54ad3bdccb5 is building: [`#2873`](https://buildkite.com/bun/bun/builds/2873)

File: src/bun.js/bindings/JSBuffer.cpp [hunk 1]
```diff
@@ -487,6 +499,10 @@ static inline JSC::EncodedJSValue jsBufferConstructorFunction_allocBody(JSC::JSG
 
     double lengthDouble = lengthValue.toIntegerWithTruncation(lexicalGlobalObject);
 
+    if (UNLIKELY(lengthDouble != lengthDouble)) {
+        throwNodeRangeError(lexicalGlobalObject, scope, "Buffer size must be 0...9007199254740991, got NaN"_s);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBuffer.cpp]: can just make throwRangeError do this instead of adding another

File: src/bun.js/bindings/JSBuffer.cpp [hunk 2]
```diff
@@ -930,16 +914,13 @@ static inline JSC::EncodedJSValue jsBufferPrototypeFunction_compareBody(JSC::JSG
...
-    }
 
-    JSC::JSUint8Array* view = JSC::jsDynamicCast<JSC::JSUint8Array*>(callFrame->uncheckedArgument(0));
+    auto arg0 = callFrame->argument(0);
+    JSC::JSUint8Array* view = JSC::jsDynamicCast<JSC::JSUint8Array*>(arg0);
 
     if (UNLIKELY(!view)) {
-        throwVMTypeError(lexicalGlobalObject, throwScope, "Expected Uint8Array"_s);
-        return JSValue::encode(jsUndefined());
+        throwScope.throwException(lexicalGlobalObject, Bun::Bun__ERR_INVALID_ARG_TYPE_static3(lexicalGlobalObject, "target"_s, "Buffer or Uint8Array"_s, arg0));
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBuffer.cpp]: we can use template <...Args pattern for this

Comment (Jarred-Sumner) [reply src/bun.js/bindings/JSBuffer.cpp]: e.g. from webkit:
```
    template <typename ...Args>
    NEVER_INLINE UnexpectedResult WARN_UNUSED_RETURN fail(Args... args) const
    {
        using namespace FailureHelper; // See ADL comment in namespace above.
        return UnexpectedResult(makeString("WebAssembly.Module doesn't parse at byte "_s, m_offset, ": "_s, makeString(args)...));
    }
    ```

File: src/bun.js/modules/NodeBufferModule.h [hunk 3]
```diff
@@ -169,15 +169,15 @@ DEFINE_NATIVE_MODULE(NodeBuffer) {
       JSC::jsNumber(MAX_ARRAY_BUFFER_SIZE));
 
   put(JSC::Identifier::fromString(vm, "kStringMaxLength"_s),
-      JSC::jsNumber(std::numeric_limits<unsigned>().max()));
```
Comment (Jarred-Sumner) [inline src/bun.js/modules/NodeBufferModule.h:172]: in JSC, the max string length is unsigned int not int

File: src/bun.js/modules/NodeBufferModule.h [hunk 4]
```diff
@@ -169,15 +169,15 @@ DEFINE_NATIVE_MODULE(NodeBuffer) {
...
+      JSC::jsNumber(std::numeric_limits<int>().max()));
 
   JSC::JSObject *constants = JSC::constructEmptyObject(
       lexicalGlobalObject, globalObject->objectPrototype(), 2);
   constants->putDirect(vm, JSC::Identifier::fromString(vm, "MAX_LENGTH"_s),
                        JSC::jsNumber(MAX_ARRAY_BUFFER_SIZE));
   constants->putDirect(vm,
                        JSC::Identifier::fromString(vm, "MAX_STRING_LENGTH"_s),
-                       JSC::jsNumber(std::numeric_limits<unsigned>().max()));
```
Comment (Jarred-Sumner) [inline src/bun.js/modules/NodeBufferModule.h:180]: in JSC, the max string length is unsigned int not int

File: test/js/node/buffer.test.js [hunk 5]
```diff
@@ -2003,9 +2003,9 @@ for (let withOverridenBufferWrite of [false, true]) {
 
       it("constants", () => {
         expect(BufferModule.constants.MAX_LENGTH).toBe(4294967296);
-        expect(BufferModule.constants.MAX_STRING_LENGTH).toBe(4294967295);
```
Comment (Jarred-Sumner) [inline test/js/node/buffer.test.js:2006]: in JSC, the max string length is unsigned int not int

File: test/js/node/buffer.test.js [hunk 6]
```diff
@@ -2003,9 +2003,9 @@ for (let withOverridenBufferWrite of [false, true]) {
...
+        expect(BufferModule.constants.MAX_STRING_LENGTH).toBe(2147483647);
         expect(BufferModule.default.constants.MAX_LENGTH).toBe(4294967296);
-        expect(BufferModule.default.constants.MAX_STRING_LENGTH).toBe(4294967295);
```
Comment (Jarred-Sumner) [inline test/js/node/buffer.test.js:2008]: in JSC, the max string length is unsigned int not int

File: src/bun.js/bindings/JSBuffer.cpp (hunk 2 above)
Comment (nektro) [reply src/bun.js/bindings/JSBuffer.cpp]: these two variants have the same exact signature but a slightly different message so its much simpler to keep them separate for now

File: src/bun.js/bindings/JSBuffer.cpp (hunk 1 above)
Comment (nektro) [reply src/bun.js/bindings/JSBuffer.cpp]: this is calling the same method with a different message, not a new function

File: src/bun.js/modules/NodeBufferModule.h [hunk 7]
```diff
@@ -46,8 +46,10 @@ JSC_DEFINE_HOST_FUNCTION(jsBufferConstructorFunction_isUtf8,
     }
 
     if (UNLIKELY(impl->isDetached())) {
-      throwTypeError(lexicalGlobalObject, throwScope,
-                     "ArrayBuffer is detached"_s);
+      throwScope.throwException(
+          lexicalGlobalObject,
+          Bun::new_ERR_INVALID_STATE(lexicalGlobalObject,
```
Comment (Jarred-Sumner) [inline src/bun.js/modules/NodeBufferModule.h]: Why not name it something more like
```
throwException(throwScope, Bun::ERR_INVALID_STATE(lexicalGlobalScope, "Cannot validate detached buffer"_s))
```

Comment (Jarred-Sumner) [reply src/bun.js/modules/NodeBufferModule.h]: or

```c++
Bun::throwInvalidState(throwScope, lexicalGlobalObject,  "Cannot validate on a detached buffer"_s)
```

File: src/bun.js/bindings/headers-handwritten.h [hunk 8]
```diff
@@ -263,6 +263,8 @@ typedef void WebSocketHTTPSClient;
 typedef void WebSocketClient;
 typedef void WebSocketClientTLS;
 
+#define MAX_STRING_LENGTH std::numeric_limits<unsigned>().max()
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/headers-handwritten.h]:
```suggestion
#define MAX_STRING_LENGTH WTF::String::MaxLength
```

File: src/js/builtins/JSBufferConstructor.ts [hunk 9]
```diff
@@ -29,6 +29,11 @@ export function from(items) {
       }
     }
   }
+  if (typeof items === "object") {
+    if (items.type === "Buffer" && Array.isArray(items.data)) {
+      return new $Buffer(Uint8Array.from(items.data).buffer);
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferConstructor.ts]: This creates two Uint8Array.

We don't have to pay that performance overhead.

```suggestion
      return new $Buffer(items.data);
```

File: src/js/builtins/JSBufferPrototype.ts [hunk 10]
```diff
@@ -317,13 +425,15 @@ export function writeUIntLE(this: BufferExt, value, offset, byteLength) {
...
-    default: {
-      throw new RangeError("byteLength must be >= 1 and <= 6");
-    }
   }
   return offset + byteLength;
 }
+
 export function writeUIntBE(this: BufferExt, value, offset, byteLength) {
+  require("internal/validators").validateIntRange(byteLength, "byteLength", 1, 6);
+  require("internal/validators").validateIntRange(offset, "offset", 0, this.byteLength - byteLength);
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferPrototype.ts]: `DataView` does it's own validation

Do we need to validate it twice?

File: src/js/builtins/JSBufferPrototype.ts [hunk 11]
```diff
@@ -353,49 +463,66 @@ export function writeUIntBE(this: BufferExt, value, offset, byteLength) {
...
-      throw new RangeError("byteLength must be >= 1 and <= 6");
-    }
   }
   return offset + byteLength;
 }
 
 export function writeFloatLE(this: BufferExt, value, offset) {
+  if (offset === undefined) offset = 0;
+  require("internal/validators").validateInteger(offset, "offset", 0, this.byteLength - this.byteOffset - 4);
+  if (this.byteOffset + offset + 4 > this.byteLength) throw require("internal/errors").ERR_BUFFER_OUT_OF_BOUNDS();
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferPrototype.ts]: Ditto here in all of these re: validation

Also, when you see us access the same property more than once, it's better to assign it to a local variable.

File: src/js/internal/validators.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
+const { ERR_INVALID_ARG_TYPE, ERR_OUT_OF_RANGE, ERR_BUFFER_OUT_OF_BOUNDS } = require("internal/errors");
+
+export default {
+  validateIntRange: function (value, name, min, max) {
```
Comment (Jarred-Sumner) [inline src/js/internal/validators.ts]: Isn't there an implementation in native code already?

File: src/js/builtins/JSBufferPrototype.ts (hunk 10 above)
Comment (nektro) [reply src/js/builtins/JSBufferPrototype.ts]: some but not all and none of the errors have codes

File: src/bun.js/bindings/JSBuffer.cpp [hunk 13]
```diff
@@ -990,39 +960,44 @@ static inline JSC::EncodedJSValue jsBufferPrototypeFunction_compareBody(JSC::JSG
     switch (callFrame->argumentCount()) {
     default:
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBuffer.cpp:961]: does this always go to the `default` value?

File: src/bun.js/bindings/JSBuffer.cpp [hunk 14]
```diff
@@ -2223,25 +2186,43 @@ static const HashTableValue JSBufferPrototypeTableValues[]
...
-          { "writeUint32BE"_s, static_cast<unsigned>(JSC::PropertyAttribute::Builtin), NoIntrinsic, { HashTableValue::BuiltinGeneratorType, jsBufferPrototypeWriteUInt32BECodeGenerator, 1 } },
-          { "writeUint32LE"_s, static_cast<unsigned>(JSC::PropertyAttribute::Builtin), NoIntrinsic, { HashTableValue::BuiltinGeneratorType, jsBufferPrototypeWriteUInt32LECodeGenerator, 1 } },
-          { "writeBigUint64BE"_s, static_cast<unsigned>(JSC::PropertyAttribute::Builtin), NoIntrinsic, { HashTableValue::BuiltinGeneratorType, jsBufferPrototypeWriteBigUInt64BECodeGenerator, 1 } },
-          { "writeBigUint64LE"_s, static_cast<unsigned>(JSC::PropertyAttribute::Builtin), NoIntrinsic, { HashTableValue::BuiltinGeneratorType, jsBufferPrototypeWriteBigUInt64LECodeGenerator, 1 } },
       };
 
+#define ALIAS(to, from)                                                                          \
+    do {                                                                                         \
+        auto original_ident = Identifier::fromString(vm, ASCIILiteral::fromLiteralUnsafe(from)); \
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBuffer.cpp:2193]:
```suggestion
        const auto original_ident = Identifier::fromString(vm, from ""_s); \
```

File: src/bun.js/bindings/JSBuffer.cpp [hunk 15]
```diff
@@ -2223,25 +2186,43 @@ static const HashTableValue JSBufferPrototypeTableValues[]
...
+        auto original = this->getDirect(vm, original_ident);                                     \
+        auto alias_ident = Identifier::fromString(vm, ASCIILiteral::fromLiteralUnsafe(to));      \
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBuffer.cpp:2195]:
```suggestion
        const auto alias_ident = Identifier::fromString(vm, to ""_s);      \
```

File: src/js/builtins/JSBufferConstructor.ts [hunk 16]
```diff
@@ -29,6 +29,11 @@ export function from(items) {
       }
     }
   }
+  if (typeof items === "object") {
+    if (items.type === "Buffer" && Array.isArray(items.data)) {
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferConstructor.ts:33]:
```suggestion
    const data = items.data;
    if (items.type === "Buffer" && Array.isArray(data)) {
```

File: src/js/builtins/JSBufferConstructor.ts [hunk 17]
```diff
@@ -29,6 +29,11 @@ export function from(items) {
...
+      return new $Buffer(items.data);
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferConstructor.ts:34]:
```suggestion
      return new $Buffer(data);
```

File: src/js/builtins/JSBufferPrototype.ts [hunk 18]
```diff
@@ -9,45 +9,79 @@ interface BufferExt extends Buffer {
 }
 
 export function setBigUint64(this: BufferExt, offset, value, le) {
+  if (offset === undefined) offset = 0;
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferPrototype.ts:12]:
```suggestion
  offset ??= 0;
```

File: src/js/builtins/JSBufferPrototype.ts [hunk 19]
```diff
@@ -9,45 +9,79 @@ interface BufferExt extends Buffer {
...
   return (this.$dataView ||= new DataView(this.buffer, this.byteOffset, this.byteLength)).setBigUint64(
     offset,
     value,
     le,
   );
 }
+
 export function readInt8(this: BufferExt, offset) {
+  if (offset === undefined) offset = 0;
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferPrototype.ts:21]:
```suggestion
  offset ??= 0;
```

File: src/js/builtins/JSBufferPrototype.ts [hunk 20]
```diff
@@ -9,45 +9,79 @@ interface BufferExt extends Buffer {
...
+  require("internal/validators").validateInteger(offset, "offset", 0);
   return (this.$dataView ||= new DataView(this.buffer, this.byteOffset, this.byteLength)).getInt8(offset);
 }
+
 export function readUInt8(this: BufferExt, offset) {
+  if (offset === undefined) offset = 0;
```
Comment (Jarred-Sumner) [inline src/js/builtins/JSBufferPrototype.ts:27]:
```suggestion
  offset ??= 0;
```

File: src/js/builtins/JSBufferPrototype.ts (hunk 18 above)
Comment (nektro) [reply src/js/builtins/JSBufferPrototype.ts:12]: `a ??= b` also assigns if `a` is `null` which i do not want to do in this case (its emulating setting a default parameter value)

File: src/js/builtins/JSBufferPrototype.ts (hunk 19 above)
Comment (nektro) [reply src/js/builtins/JSBufferPrototype.ts:21]: see above

File: src/js/builtins/JSBufferPrototype.ts (hunk 20 above)
Comment (nektro) [reply src/js/builtins/JSBufferPrototype.ts:27]: see above

File: src/bun.js/bindings/JSBuffer.cpp (hunk 13 above)
Comment (nektro) [reply src/bun.js/bindings/JSBuffer.cpp:961]: only then there's 5 or more arguments

File: src/bun.js/bindings/JSBuffer.cpp (hunk 14 above)
Comment (nektro) [reply src/bun.js/bindings/JSBuffer.cpp:2193]: `_s` calls `ASCIILiteral::fromLiteralUnsafe` though so doesn't this add a step?

File: src/bun.js/bindings/JSBuffer.cpp [hunk 21]
```diff
@@ -2223,25 +2186,43 @@ static const HashTableValue JSBufferPrototypeTableValues[]
...
+        this->putDirect(vm, alias_ident, original, PropertyAttribute::Builtin | 0);              \
+    } while (false);
+
 void JSBufferPrototype::finishCreation(VM& vm, JSC::JSGlobalObject* globalThis)
 {
     Base::finishCreation(vm);
     JSC_TO_STRING_TAG_WITHOUT_TRANSITION();
     reifyStaticProperties(vm, JSBuffer::info(), JSBufferPrototypeTableValues, *this);
+
+    ALIAS("readUintBE", "readUIntBE");
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/JSBuffer.cpp:2205]: We should update the static hashtables to support this as a PropertyAttribute or something

**Pull request merged**